src/io/ipc/read/deserialize.rs

```diff
@@ -80,51 +105,162 @@ fn read_buffer<T: NativeType, R: Read + Seek>(
             reader.read_exact(slice)?;
         }
     } else {
-        // slow case where we must reverse bits
-        let mut slice = vec![0u8; bytes];
-        reader.read_exact(&mut slice)?;
-
-        if !is_little_endian {
-            let chunks = slice.chunks_exact(std::mem::size_of::<T>());
-            buffer
-                .as_mut_slice()
-                .iter_mut()
-                .zip(chunks)
-                .try_for_each(|(slot, chunk)| {
-                    let a: T::Bytes = match chunk.try_into() {
-                        Ok(a) => a,
-                        Err(_) => unreachable!(),
-                    };
-                    *slot = T::from_be_bytes(a);
-                    ArrowResult::Ok(())
-                })
-                .unwrap();
-        }
+        read_big_endian(reader, bytes, &mut buffer, is_little_endian)?;
+    }
+    Ok(buffer)
+}
+
+fn read_compressed_buffer<T: NativeType, R: Read + Seek>(
+    reader: &mut R,
+    buffer_length: usize,
+    length: usize,
+    is_little_endian: bool,
+    compression: BodyCompression,
+) -> Result<MutableBuffer<T>> {
+    if is_little_endian != is_native_little_endian() {
+        return Err(ArrowError::NotYetImplemented(
+            "Reading compressed and big endian IPC".to_string(),
+        ));
     }
 
-    Ok(buffer.into())
+    // it is undefined behavior to call read_exact on un-initialized, https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
+    // see also https://github.com/MaikKlein/ash/issues/354#issue-781730580
+    let mut buffer = MutableBuffer::<T>::from_len_zeroed(length);
+
+    // decompress first
+    // todo: move this allocation to an external buffer for re-use
+    let mut slice = vec![0u8; buffer_length];
+    reader.read_exact(&mut slice)?;
+
+    match compression.codec() {
+        CompressionType::LZ4_FRAME => {
+            // fast case where we can just copy the contents as is
+            unsafe {
+                // transmute T to bytes.
+                let out_slice = std::slice::from_raw_parts_mut(
+                    buffer.as_mut_ptr() as *mut u8,
+                    length * std::mem::size_of::<T>(),
+                );
+                compression::decompress_lz4(&slice[8..], out_slice)?
```

Comment:
If a writer tries to compress a buffer, and then it finds that the compressed buffer ratio isn't good enough, it can still write uncompressed buffers. In that case, we have to check if the first 8 bytes = `-1i64`. It's some of the details that I struggled with

Reply:
I am following the c++ implementation [here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/ipc/reader.cc#L399), but I can't find that case there :/

Reply:
I also can't find it now, but I'll look in more detail; as I recall a review comment being raised in the Java implementation for this detal.

Reply:
Wouldn't it be fine for now to use the same as the c++ implementation does, and improve that behavior over a future PR? None of this results in UB; the worse case is a panic.

Reply:
I agree, it definitely would :)